File: 2016/02/02.py
def generator(input):
    return input.splitlines()

def part_1(procedure):
    position = 5 
    code     = ''
    for number in procedure: 
        for instruction in number: 
            match instruction:
                case 'D':
                    if position not in [7, 8, 9]: position += 3
                case 'U': 
                    if position not in [1, 2, 3]: position -= 3
                case 'L':
                    if position not in [1, 4, 7]: position -= 1
                case 'R':
                    if position not in [3, 6, 9]: position += 1
        code += format(position) 
    return code

def part_2(procedure):
    position = 5
    code     = ''
    for number in procedure: 
        for instruction in number: 
            match instruction:
                case 'D': 
                    if   position in [1,11]: position += 2
                    elif position in [2,3,4,6,7,8]: position += 4     
                case 'U': 
                    if   position in [3,13]: position-=2
                    elif position in [6,7,8,10,11,12]: position -= 4
                case 'L':
                    if   position not in [1, 2, 5, 10, 13]: position -= 1
                case 'R':
                    if   position not in [1, 4, 9, 12, 13]: position += 1
        code += format(position,'X')
    return code

File: 2016/02/02.rs
aoc::main!();

fn parser(input: &str) -> Vec<&str> {
    input
    .lines()
    .collect_vec()
 }

 fn part_1(procedure: Vec<&str>) -> String {
    solver(procedure.clone(), false)
    .into_iter()
    .map(|(x,y) :(i32, i32)| (1 + 3 * y + x).to_string())
    .join("")
    }
    
fn part_2(procedure: Vec<&str>) -> String {
    solver(procedure.clone(), true)
    .into_iter()
    .map(|(x,y) :(i32, i32)| {format!("{:X}", 1 + y * (y - ((y == 4) as i32)) - (2 - y).abs() + x)})
    .join("")
}


fn solver(procedure: Vec<&str>, bathroom_keyboard: bool) -> Vec<(i32, i32)> {
    procedure
    .iter()
    .map(|instr|{
        instr
        .chars()
        .fold((1,1), |acc : (i32, i32), c| {
            let bounds = 
                match bathroom_keyboard { 
                    false => (0, 2, 2, 0),
                    true => ((acc.0 - 2).abs(), 4 - (acc.0 - 2).abs(), 4 - (acc.1 - 2).abs(), (acc.1 - 2).abs())};
            match c {
                'U' => (acc.0, (bounds.0).max(acc.1 - 1)),
                'D' => (acc.0, (bounds.1).min(acc.1 + 1)),
                'R' => ((bounds.2).min(acc.0 + 1), acc.1),
                'L' => ((bounds.3).max(acc.0 - 1), acc.1),
                _ => unreachable!()}})})
    .collect_vec()
}
